libcaf_net/test/net-test.hpp

```diff
@@ -105,6 +106,63 @@ class mock_stream_transport : public caf::net::octet_stream::lower_layer {
   caf::net::multiplexer* mpx_;
 };
 
+class mock_web_socket_app : public caf::net::web_socket::upper_layer::server {
+public:
+  // -- constructor ------------------------------------------------------------
+
+  mock_web_socket_app(bool behave_as_server);
+
+  // -- factories --------------------------------------------------------------
+
+  static auto make(bool behave_as_server = false) {
+    return std::make_unique<mock_web_socket_app>(behave_as_server);
+  }
+
+  // -- initialization ---------------------------------------------------------
+
+  caf::error start(caf::net::web_socket::lower_layer* ll) override;
+
+  // -- implementation ---------------------------------------------------------
+
+  caf::error accept(const caf::net::http::request_header& hdr) override;
+
+  void prepare_send() override {
+    // nop
+  }
```

Comment:
Even if they are trivial, our convention is to put virtual functions always into the `.cpp` file (except for templates, of course).